axiom/datasets.go

```diff
@@ -379,70 +385,109 @@ func (s *DatasetsService) IngestEvents(ctx context.Context, id string, events []
 // Restrictions for field names (JSON object keys) can be reviewed here:
 // https://www.axiom.co/docs/usage/field-restrictions.
 //
-// As this method keeps a connection open until the channel is closed, it is not
-// advised to use this method for long-running ingestions.
+// Events are ingested in batches. A batch is either 1024 events for unbuffered
+// channels or the capacity of the channel for buffered channels. A batch is
+// sent to the server as soon as it is full or after one second or when the
+// channel is closed.
 func (s *DatasetsService) IngestChannel(ctx context.Context, id string, events <-chan Event, options ...ingest.Option) (*ingest.Status, error) {
+	const (
+		defaultBatchSize    = 1024
+		defaultSendInterval = time.Second
+	)
+
 	ctx, span := s.client.trace(ctx, "Datasets.IngestChannel", trace.WithAttributes(
 		attribute.String("axiom.dataset_id", id),
 		attribute.Int("axiom.channel.capacity", cap(events)),
 	))
 	defer span.End()
 
-	// Apply supplied options.
-	var opts ingest.Options
-	for _, option := range options {
-		option(&opts)
-	}
+	var (
+		errCh        = make(chan error)
+		ingestStatus ingest.Status
+	)
+	go func() {
+		defer close(errCh)
 
-	path, err := AddOptions(s.basePath+"/"+id+"/ingest", opts)
-	if err != nil {
-		return nil, spanError(span, err)
-	}
+		// Batch is either 1024 events for unbuffered channels or the capacity
+		// of the channel for buffered channels.
+		batchSize := defaultBatchSize
+		if cap(events) != 0 {
+			batchSize = cap(events)
+		}
+		batch := make([]Event, 0, batchSize)
 
-	pr, pw := io.Pipe()
+		ingestBatch := func(ctx context.Context) error {
+			if len(batch) == 0 {
+				return nil
+			}
 
-	zsw, err := zstd.NewWriter(pw)
-	if err != nil {
-		_ = pr.Close()
-		_ = pw.Close()
-		return nil, err
-	}
+			// Reset batch buffer after attempted ingest, regardless of the
+			// outcome.
+			defer func() {
+				batch = make([]Event, 0, batchSize)
```

Comment:
We could trim the slice instead (or do we have to create a copy?)

```suggestion
				batch = batch[:0]
```

Reply:
But if the slice is extended again, the original data reappears, no? Which shouldn't be a problem as we don't extend it.